lib/c-api/src/wasm_c_api/unstable/middlewares/mod.rs

```diff
@@ -0,0 +1,40 @@
+//! Unstable non-standard Wasmer-specific types to manipulate module
+//! middlewares.
+
+pub mod metering;
+
+use super::super::engine::wasm_config_t;
+use std::sync::Arc;
+use wasmer::ModuleMiddleware;
+
+#[cfg(all(feature = "middlewares", not(feature = "compiler")))]
+compile_error!("The `middlewares` feature requires the `compiler` feature to be turned on");
+
+/// Opaque representing any kind of middleware.
+///
+/// Used by `wasm_config_push_middleware`. A specific middleware is
+/// transformed into this type to get a generic middleware. See for
+/// example
+/// [`wasmer_metering_as_middleware`][metering::wasmer_metering_as_middleware].
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+pub struct wasmer_middleware_t {
+    pub(in crate::wasm_c_api) inner: Arc<dyn ModuleMiddleware>,
+}
+
+/// Updates the configuration to add a module middleware.
+///
+/// This function takes ownership of `middleware`.
+///
+/// This is a Wasmer-specific function.
+///
+/// # Example
+///
+/// See the documentation of the [`metering`] module.
+#[no_mangle]
+pub extern "C" fn wasm_config_push_middleware(
```

Comment:
I'd assume middleware happens at a higher level than `wasm_config`, as `wasm_config` is used for setting up the initial engine / store. Not a huge deal, but it means that it's harder to share a store with modules using different middlewares. (I'm just assuming we allow a more flexible usage in the Rust API)